cmd/release-controller/http_candidate.go

```diff
@@ -0,0 +1,336 @@
+	border-width: 6px;
+}
+.upgrade-track-line.start {
+	top: 18px;
+	height: 31px;
+	display: block;
+}
+.upgrade-track-line.middle {
+	display: block;
+}
+.upgrade-track-line.end {
+	top: -1px;
+	height: 16px;
+	display: block;
+}
+td.upgrade-track {
+	width: 16px;
+	position: relative;
+	padding-left: 2px;
+	padding-right: 2px;
+}
+</style>
+<div class="row">
+<div class="col">
+	<table class="table text-nowrap">
+		<thead>
+			<tr>
+				<th title="Candidate tag for next release">Name</th>
+				<th title="Tag(s) of release this can upgrade FROM">Upgrades</th>
+				<th title="Creation time">Creation time</th>
+			</tr>
+		</thead>
+		<tbody>
+		{{ range $candidate := $list.Items }}
+			<tr>
+				<td> <a href="/releasestream/releasetag/{{ $candidate.FromTag }}" >{{ $candidate.FromTag }} </a></td>
+				<td>{{ range $prev := $candidate.UpgradeFrom }}
+					<a href="/releasestream/{{ $prev }}/release/{{ $prev }}"> {{ $prev }} </a>, 
+					{{ end }}
+				</td>
+				<td>{{ $candidate.CreationTime }}</td>
+			</tr>
+		{{ end }}
+		</tbody>
+	</table>
+</div>
+</div>
+{{ end }}
+`
+
+func (c *Controller) httpReleaseCandidateList(w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	defer func() { glog.V(4).Infof("rendered in %s", time.Now().Sub(start)) }()
+	vars := mux.Vars(req)
+	releaseStreamName := vars["release"]
+	releaseCandidateList, err := c.findReleaseCandidates(releaseStreamName)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	switch req.URL.Query().Get("format") {
+	case "json":
+		data, err := json.MarshalIndent(&releaseCandidateList, "", "  ")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		fmt.Fprintf(w, string(data))
+	default:
+		fmt.Fprintf(w, htmlPageStart, "Release Status")
+		page := template.Must(template.New("candidatePage").Funcs(
+			template.FuncMap{
+				"nextReleaseName": func(list *ReleaseCandidateList) string {
+					if list == nil || list.Items == nil || len(list.Items) == 0 {
+						return "next release"
+					}
+					return list.Items[0].Name
+				},
+			},
+		).Parse(candidatePageHtml))
+
+		if err := page.Execute(w, releaseCandidateList); err != nil {
+			glog.Errorf("Unable to render page: %v", err)
+		}
+		fmt.Fprintln(w, htmlPageEnd)
+	}
+}
+
+func (c *Controller) httpReleaseCandidate(w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	var successPercent float64
+	defer func() { glog.V(4).Infof("rendered in %s", time.Now().Sub(start)) }()
+	vars := mux.Vars(req)
+	releaseStreamName := vars["release"]
+	releaseCandidateList, err := c.findReleaseCandidates(releaseStreamName)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	switch req.URL.Query().Get("format") {
+	default:
+		var candidate *ReleaseCandidate
+		if releaseCandidateList[releaseStreamName] != nil && len(releaseCandidateList[releaseStreamName].Items) > 0 {
```

Comment:
This isn't really the described logic.  You only get one releaseStreamName - you should pull this if statement into two separate if statements (one up above the switch that returns 404 if stream isn't found). 

The requirement is that you return an error if there are no candidates.  The appropriate response code for that is probably 204 no content, but we should return a json body that is distinct from "I got a candidate".

Generally it helps to describe out the JSON you should be returning, so to satisfy this requirement the no candidate case is probably:

```
{
  "candidate": null
}
```

The JSON returned when there is a candidate should be:

```
{
  "candidate": { 
     ... // info for end user
  }
}
```

Since JSON is dependent on format, it's best to structure this code as a struct that has the single field `*ReleaseCandidate`, then have your if block for Items include it.